src/core/ir/discriminator.rs

```diff
@@ -0,0 +1,1121 @@
+use std::collections::HashSet;
+use std::fmt::Write;
+use std::ops::{Deref, DerefMut};
+
+use anyhow::{bail, Result};
+use async_graphql::Value;
+use indenter::indented;
+use indexmap::IndexMap;
+
+use crate::core::config::Type;
+use crate::core::valid::{Cause, Valid, Validator};
+
+/// BitVecesents the type name for the resolved value.
+/// It is used when the GraphQL executor needs to resolve values of a union
+/// type. In order to select the correct fields, the executor must know the
+/// exact type name for each resolved value. And in cases when the output is
+/// list of union type it should resolve exact type for every entry in list.
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum TypeName {
+    Single(String),
+    Vec(Vec<String>),
+}
```

Comment:
How about storing a ref of string?

Reply:
The unconformable part is that async_graphql API requires 'static str or String to pass https://docs.rs/async-graphql/latest/async_graphql/dynamic/struct.FieldValue.html#method.with_type

I've put an experimental implementation as separate pr for this using `String::leak` here https://github.com/tailcallhq/tailcall/pull/2210